### ISSUE TITLE
Futex timeout is not virtualized for time namespaces (Java JAR poor performance after restore)

### ISSUE THREAD

**Issue by HansvdLaan** [checkpoint-restore/criu#1232]
Hey,

I'm trying out CRIU in combination with Java. I'm trying to dump and restore a simple java counter which saves a number to a file. When I dump a simple dummy java program and subsequently restart it, the restored java program uses almost a full core and has become extremely slow. 

Since it's just a simple program, it seems as I forgot some parameter. Does anyone see what went wrong?

```
java -jar ./counter.jar 3000
criu dump -t $(pidof java) -o dump.log --shell-job
criu restore -o restore.log --shell-job
```

The java program:

```java
public class Fun {
    public static void main(String[] args) throws InterruptedException {
        for(int i = 0; i < Integer.valueOf(args[0]); i++) {
            try
            {
                File file = new File("test.log");
                FileWriter fw = new FileWriter(file,true); //the true will append the new data
                fw.write("Counter: " + i + "\n");//appends the string to the file
                fw.close();
            }
            catch(IOException ioe) {
                System.err.println("IOException: " + ioe.getMessage());
            }
            Thread.sleep(1000);
        }
    }
}
```

An strace:
```
[...]
[pid 42558] <... futex resumed>)        = 0
[pid 42540] futex(0x7f5314009b28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 42558] futex(0x7f5314215a78, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=27621, tv_nsec=904997991}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
[pid 42549] futex(0x7f53141c0e28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 42558] <... futex resumed>)        = -1 ETIMEDOUT (Connection timed out)
[pid 42540] <... futex resumed>)        = 0
[pid 42558] futex(0x7f5314215a28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 42549] <... futex resumed>)        = 0
[pid 42558] <... futex resumed>)        = 0
[pid 42540] futex(0x7f5314009b78, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=27622, tv_nsec=270025790}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
[pid 42558] futex(0x7f5314215a78, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=27621, tv_nsec=905041162}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
[pid 42549] futex(0x7f53141c0e78, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=27622, tv_nsec=668036706}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
[pid 42558] <... futex resumed>)        = -1 ETIMEDOUT (Connection timed out)
[pid 42540] <... futex resumed>)        = -1 ETIMEDOUT (Connection timed out)
[pid 42558] futex(0x7f5314215a28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 42549] <... futex resumed>)        = -1 ETIMEDOUT (Connection timed out)
[pid 42558] <... futex resumed>)        = 0
[pid 42540] futex(0x7f5314009b28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 42558] futex(0x7f5314215a78, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=27621, tv_nsec=905091933}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
[pid 42549] futex(0x7f53141c0e28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 42540] <... futex resumed>)        = 0
[pid 42558] <... futex resumed>)        = -1 ETIMEDOUT (Connection timed out)
[pid 42549] <... futex resumed>)        = 0
[pid 42558] futex(0x7f5314215a28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 42540] futex(0x7f5314009b78, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=27622, tv_nsec=270115121}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
[pid 42558] <... futex resumed>)        = 0
[pid 42549] futex(0x7f53141c0e78, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=27622, tv_nsec=668124916}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
[pid 42558] futex(0x7f5314215a78, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=27621, tv_nsec=905143754}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
[pid 42540] <... futex resumed>)        = -1 ETIMEDOUT (Connection timed out)
[pid 42558] <... futex resumed>)        = -1 ETIMEDOUT (Connection timed out)
[pid 42549] <... futex resumed>)        = -1 ETIMEDOUT (Connection timed out)
[pid 42558] futex(0x7f5314215a28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 42540] futex(0x7f5314009b28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 42558] <... futex resumed>)        = 0
[pid 42549] futex(0x7f53141c0e28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 42558] futex(0x7f5314215a78, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=27621, tv_nsec=905191615}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
[pid 42540] <... futex resumed>)        = 0
[pid 42549] <... futex resumed>)        = 0
[pid 42558] <... futex resumed>)        = -1 ETIMEDOUT (Connection timed out)
[pid 42540] futex(0x7f5314009b78, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=27622, tv_nsec=270210680}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
[pid 42558] futex(0x7f5314215a28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 42549] futex(0x7f53141c0e78, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=27622, tv_nsec=668213658}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
[pid 42558] <... futex resumed>)        = 0
[pid 42540] <... futex resumed>)        = -1 ETIMEDOUT (Connection timed out)
[pid 42558] futex(0x7f5314215a78, FUTEX_WAIT_BITSET_PRIVATE, 0, {tv_sec=27621, tv_nsec=905246591}, FUTEX_BITSET_MATCH_ANY <unfinished ...>
[pid 42549] <... futex resumed>)        = -1 ETIMEDOUT (Connection timed out)
[pid 42558] <... futex resumed>)        = -1 ETIMEDOUT (Connection timed out)
[pid 42540] futex(0x7f5314009b28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 42558] futex(0x7f5314215a28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 42549] futex(0x7f53141c0e28, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 42558] <... futex resumed>)        = 0
[...]
```

I've tried it on Arch Linux on a physical machine with OpenJDK 1.8.0_265 and a virtual machine running Debian unstable
with both the latest version of OpenJDK 11 (11.0.8 2020-07-14) and Oracle JDK 11.0.8 (build 11.0.8+10-LTS) . I've got the latest version of CRIU.

**Comment by adrianreber** [comment]
Very interesting report. I am using Podman to checkpoint and restore a Wildfly container and sometimes I see the JVM also running at 100% on multiple cores and I assumed that it was related to wrongly restoring the container. I never tried it outside of a container, but I will try to reproduce the problem without containers using your code.

It feels like this is a new behaviour, but I am not sure. Let me try it locally to confirm.

**Comment by adrianreber** [comment]
You mention `-jar`. Is that important? I do not see any problems using following JDKs:
```
#️  java -version
openjdk version "1.8.0_265"
OpenJDK Runtime Environment (build 1.8.0_265-b01)
OpenJDK 64-Bit Server VM (build 25.265-b01, mixed mode)
#️  /usr/lib/jvm/java-11-openjdk-11.0.8.10-6.el8.x86_64/bin/java --version
openjdk 11.0.8 2020-07-14 LTS
OpenJDK Runtime Environment 18.9 (build 11.0.8+10-LTS)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.8+10-LTS, mixed mode, sharing)
```

I just took your example, added the necessary imports and ran it. Can you tell me how to run that example using `-jar` if that is important to reproduce this? It uses about 1% before and after the restore. I am testing on RHEL 8.

**Comment by HansvdLaan** [comment]
Hey,

I'm just running it as a jar since it's a use-case I probably need for our real application later. 

It does not matter if we run the main class directly or if we unpack .jar and then run it directly. We still have the performance issue . e.g., like:
```
java -cp counter.jar some.Fun 1234
``` 
or
```
unzip counter.jar 
java -cp ./some.Fun 1234
```

> Can you tell me how to run that example using -jar

I'm not sure what you mean exactly what you mean. I just run the jar with ```java -jar ./counter.jar 3000```.
If it helps, this is the maven file with which I build the jar

```maven
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>org.example</groupId>
    <artifactId>ConsoleCounter</artifactId>
    <version>1.0-SNAPSHOT</version>

    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-jar-plugin</artifactId>
                <configuration>
                    <archive>
                        <manifest>
                            <mainClass>some.Fun</mainClass>
                        </manifest>
                    </archive>
                </configuration>
            </plugin>
        </plugins>
    </build>

</project>
```


Kind regards,

Hans

**Comment by adrianreber** [comment]
Yes, I was asking how to create a jar as I am not familiar enough with Java to create one. I will not install maven, I was hoping for a one-line example, not sure if that is possible. But as having it in a jar does not seem to be important, let's just skip it.

Which kernel version do you have?

**Comment by adrianreber** [comment]
This seems to be related to the kernel.If I do a checkpoint/restore on a RHEL 8 kernel it works as expected, if I do a checkpoint and restore on a Fedora based 5.9 rc1 kernel I see that it uses all four cores and each with 100%.

It would be good to have some insights from the JVM side why a restored JVM uses so much CPU on newer kernels. Any ideas who to ask?

**Comment by hberntsen** [comment]
I've tested it on Arch Linux using 5.8.14-zen1-1-zen. @HansvdLaan and I will try an older kernel tomorrow (we're in CET so our working day ended). 

I find it interesting to see that both the openjdk and Oracle JVM face the same issue. I guess we could report the Oracle issue here: https://bugreport.java.com/bugreport/ . For OpenJDK, we could post it to the mailing list ( https://stackoverflow.com/questions/29379425/where-to-report-issues-of-openjdk-when-youre-not-a-openjdk-developer )

**Comment by adrianreber** [comment]
I went back to a 5.6 kernel and still see the problem but with a 5.3 kernel the problem is gone. So between 5.3 and 5.6 something changed in the kernel which makes restored Java processes use 100% CPU.

**Comment by hberntsen** [comment]
I've just tested it on an LTS kernel (5.4.70-1-lts) and it works fine.

**Comment by avagin** [comment]
This is due to time namespaces. I will look at this tomorrow. Thank you for the report.

**Comment by avagin** [comment]
I figured out where the problem is. We forgot to apply timens offsets in the futex syscall. We didn't know that it can use absolute timeouts. We will need to patch the kernel to fix this issue.

As a workaround, you can temporary disable the time namespace in criu:
diff --git a/criu/kerndat.c b/criu/kerndat.c
index b2c47c532..b96ec0a00 100644
--- a/criu/kerndat.c
+++ b/criu/kerndat.c
@@ -791,6 +791,7 @@ out:
 
 static int has_time_namespace(void)
 {
+       return 0;
        if (access("/proc/self/timens_offsets", F_OK) < 0) {
                if (errno == ENOENT) {
                        pr_debug("Time namespaces are not supported.\n");

**Comment by avagin** [comment]
This is the kernel fix: https://lkml.org/lkml/2020/10/15/582

**Comment by adrianreber** [comment]
I can confirm it is fixed. Just updated the kernel on my Fedora system and a restored Java container does not use more CPU than before.

@HansvdLaan can we close this?

**Comment by hberntsen** [comment]
Looks like the fix indeed landed in v5.10-rc1: https://github.com/torvalds/linux/commit/c2f7d08cccf4af2ce6992feaabb9e68e4ae0bff3#diff-a0e09fac216f72474730fca968820dba5f2b4c2155797f5008cc5a0df70cbcf2

**Comment by HansvdLaan** [comment]
Yes. To be sure, me and @hberntsen also verified. We build the latest release candidate v5.10-rc4 for debian. It works!  🎉

**Comment by mbien** [comment]
5.9.8 has the fix too; (early 5.9 kernels had the same problem)
